core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java

```diff
@@ -148,22 +155,29 @@ private void needAuthentication(ContainerRequestContext request, Optional<TokenP
         Optional<String> refreshToken = tokenPair.flatMap(TokenPair::getRefreshToken);
         if (refreshToken.isPresent()) {
             try {
-                OAuth2Client.Response response = client.refreshTokens(refreshToken.get());
-                String serializedToken = tokenPairSerializer.serialize(TokenPair.fromOAuth2Response(response));
-                request.abortWith(Response.seeOther(request.getUriInfo().getRequestUri())
-                        .cookie(OAuthWebUiCookie.create(serializedToken, tokenExpiration.map(expiration -> Instant.now().plus(expiration)).orElse(response.getExpiration())))
-                        .build());
+                redirectForNewToken(request, refreshToken.get());
             }
             catch (ChallengeFailedException e) {
                 LOG.debug(e, "Tokens refresh challenge has failed");
-                startOAuth2Challenge(request);
+                sendWwwAuthenticate(request, "Refresh token flow failed", ImmutableSet.of(TRINO_FORM_LOGIN));
             }
+            return;
```

Comment:
no, because this method is a void one (as we modify the response in Filter, rather than creating and returning it). So I need to return somewhere, to stop further processing. Still I will change it to if-else as it seems clearer than this return;